File: uva online judge/00423 - MPI Maelstrom.py
import math 
def read_test_case():
    n = int(input())
    adj_matrix = [[math.inf] * n for i in range(n)]
    for i in range(1,n):
        weights = input().split()
        for j in range(i):
            if weights[j] != 'x':
                adj_matrix[i][j] = int(weights[j])
                adj_matrix[j][i] = int(weights[j])
    return n, adj_matrix

def floyd_warshall(n, adj_matrix):
    for k in range(n):
        for i in range(n):
            for j in range(n):
                adj_matrix[i][j] = min(adj_matrix[i][j], adj_matrix[i][k] + adj_matrix[k][j])
    return adj_matrix

if __name__ == '__main__':
    n, adj_matrix = read_test_case()
    floyd_warshall(n, adj_matrix)
    print(max(adj_matrix[i][0] for i in range(1,n)))

File: uva online judge/00423 - MPI Maelstrom.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>
#define MAX_NODES 100
#define INF INT_MAX


int adj_matrix[MAX_NODES][MAX_NODES];

void read_test_case(int *n, int adj_matrix[][MAX_NODES]){
    int i, j;
    char buffer[1024];
    scanf("%d", n);
    for(i = 0; i < *n ; i++)
        for(j = 0; j < *n; j++)
            adj_matrix[i][j] = INF;
    for(i = 1; i < *n; i++){
        for(j = 0; j < i; j++){
            scanf("%s", buffer);
            if(buffer[0] != 'x'){
                adj_matrix[i][j] = atoi(buffer);
                adj_matrix[j][i] = adj_matrix[i][j];
            }
        }
    }
}

void floyd_warshall(int n, int adj_matrix[][MAX_NODES]){
    int i, j, k;
    for(k = 0; k < n; k++)
        for(i = 0; i < n; i++)
            for(j = 0; j < n; j++)
                if(adj_matrix[i][k] != INF && adj_matrix[k][j] != INF && (adj_matrix[i][k] + adj_matrix[k][j]) <  adj_matrix[i][j])
                    adj_matrix[i][j] = adj_matrix[i][k] + adj_matrix[k][j];
}

int main(){
    int n, i, max_value = 0;
    read_test_case(&n,adj_matrix);
    floyd_warshall(n, adj_matrix);
    for(i = 1; i < n; i++)
        max_value =  max_value < adj_matrix[i][0] ? adj_matrix[i][0] : max_value;
    printf("%d\n", max_value);
    return 0;
}